### PR TITLE
feat(push-templates): rating template deferred submit, custom icons, and configurable confirm button

### PR DESCRIPTION
**Pull request by deeksha-rgb**
## What's changed

Adds new capabilities to the pt_rating push template:

**Deferred submit flow**
- Star tap now shows a confirm button instead of immediately firing the rating event
- Set `pt_rating_auto_submit: "false"` to opt into deferred mode (default is auto-submit for backward compatibility)
- Confirm button tap fires the CleverTap rating event and handles the deep link

**Custom icons**
- Replace default stars with any image using `pt_icon_1` through `pt_icon_5`
- Same URL for all icons = cumulative fill (star-like behavior)
- Different URLs per icon = single-select (emoji-style, only tapped icon highlighted)
- Fewer than 5 icons supported via `pt_rating_icon_count` (min 2, max 5)

**Configurable confirm button**
- Custom text via `pt_rating_confirm_text`
- Solid, linear gradient, or radial gradient background via `pt_btn_style`
- Custom text color, border color, border width, border radius
- Centered/narrow button via `pt_rating_btn_margin_h` (dp padding each side)

## Payload keys added
| Key | Type | Description |
|-----|------|-------------|
| `pt_rating_auto_submit` | `"true"` / `"false"` | Default `"true"` |
| `pt_rating_icon_count` | `"2"`–`"5"` | Number of icons |
| `pt_icon_N` | URL | Icon image for slot N |
| `pt_icon_N_sel` | URL | Selected state image for slot N |
| `pt_rating_confirm_text` | String | Button label |
| `pt_rating_btn_margin_h` | dp | Horizontal padding to center button |
| `pt_btn_style` | `solid` / `gradient_linear` / `gradient_radial` | |
| `pt_btn_clr` | hex | Solid background color |
| `pt_btn_grad_clr1`, `pt_btn_grad_clr2` | hex | Gradient colors |
| `pt_btn_text_clr` | hex | Button text color |
| `pt_btn_border_clr` | hex | Border color |
| `pt_btn_border_width` | float | Border width (dp) |
| `pt_btn_border_radius` | float | Corner radius (dp) |

## Summary by CodeRabbit

- **New Features**
  - Added configurable image borders and rounded corners across push notification templates, including GIF support and dark-mode styling.
  - Added customizable rating icons, icon counts, grayscale selected states, confirmation text, and button styling.
  - Added improved image scaling for bordered media.

- **Bug Fixes**
  - Improved carousel navigation safety and image-key detection.
  - Fixed image layout spacing and handling of invalid styling values.

- **Documentation**
  - Documented image border, corner-radius, rating, and template configuration options.
  - Added Version 2.5.0 release notes.

### CODE REVIEW

**Comment by deeksha-rgb** [comment]
@coderabbitai review

**Comment by shubhambjadhavar** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/clevertap-platform/pr-checks/baa6da0e-c0f3-4fb2-90be-3cbb3d297ae3?refs=1ea3d54193610ed5e604fe50df632db877e58f5d&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/clevertap-platform/pr-checks/baa6da0e-c0f3-4fb2-90be-3cbb3d297ae3/license?refs=1ea3d54193610ed5e604fe50df632db877e58f5d&source=prChecksComment) |
| :white_check_mark: | Code Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/clevertap-platform/project/9a1fd2dd-3229-4564-b8c3-9660985efc7a/pr-check/80081bf6-ef1f-4572-9ea1-7260dd04b1db?refs=1ea3d54193610ed5e604fe50df632db877e58f5d&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/PushTemplateReceiver.java [hunk 1]
```diff
@@ -408,6 +418,10 @@ private void handleRatingNotification(Context context, Bundle extras, Intent int
                 return;
             }
 
+            if (deepLinkList.isEmpty()) {
+                PTLog.verbose("Rating: deepLinkList is empty, cannot handle star tap");
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (deeksha-rgb) [reply clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/PushTemplateReceiver.java]: Fixed in commit 63008fbcd. Removed the early return. Deep links are now optional — star rendering, confirm button, and event firing all work without any pt_dl_* keys.

File: clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/PushTemplateReceiver.java [hunk 2]
```diff
@@ -457,50 +471,134 @@ private void handleRatingNotification(Context context, Bundle extras, Intent int
...
+                }
             }
             cancelRatingClickIntents(context,intent);
             extras.putString(Constants.DEEP_LINK_KEY, pt_dl_clicked);
-            contentViewRating.setOnClickPendingIntent(R.id.tVRatingConfirmation,
-                    LaunchPendingIntentFactory.getActivityIntent(extras, context));
+
+            // Default is auto-submit; deferred submit is opt-in via pt_rating_auto_submit=false
+            String autoSubmitVal = extras.getString(PTConstants.PT_RATING_AUTO_SUBMIT);
+            boolean autoSubmit = autoSubmitVal == null || !"false".equalsIgnoreCase(autoSubmitVal);
```
[the remark this answers is not in the document]
Comment (deeksha-rgb) [reply clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/PushTemplateReceiver.java]: Fixed in commit 63008fbcd. Now checks getString() first; if null, falls back to getBoolean() so both representations are handled

File: clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/PushTemplateReceiver.java [hunk 3]
```diff
@@ -521,11 +619,12 @@ private void handleRatingNotification(Context context, Bundle extras, Intent int
                 notification = notificationBuilder.build();
 
                 notificationManager.notify(notificationId, notification);
+                for (Bitmap b : bitmapsToRecycle) { if (!b.isRecycled()) b.recycle(); }
             }
```
[the remark this answers is not in the document]
Comment (deeksha-rgb) [reply clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/PushTemplateReceiver.java:625]: Fixed in commit 63008fbcd. Only derived bitmaps created by toGreyscale() are recycled after notify(). Source bitmaps from getImageBitmap() are no longer touched.

File: clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/TemplateDataFactory.kt [hunk 4]
```diff
@@ -450,11 +478,16 @@ internal object TemplateDataFactory {
...
+            imageBorderData = defaultMediaData.imageBorderData
         )
     }
 
-    private fun createCollapsedMediaDataWithoutFallback(extras: Bundle, defaultAltText: String): MediaData? {
+    private fun createCollapsedMediaDataWithoutFallback(
+        extras: Bundle,
+        defaultAltText: String,
+        imageBorderData: ImageBorderData = ImageBorderData()
+    ): MediaData? {
```
[the remark this answers is not in the document]
Comment (deeksha-rgb) [reply clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/TemplateDataFactory.kt:490]: This is unrelated to the rating template changes in this PR. Will address in a separate PR for the vertical image template.

File: clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/Utils.java [hunk 5]
```diff
@@ -600,6 +603,17 @@ public static int getColour(String clr, String default_clr) {
...
+    public static Bitmap toGreyscale(Bitmap src) {
+        Bitmap result = Bitmap.createBitmap(src.getWidth(), src.getHeight(), Bitmap.Config.ARGB_8888);
+        Canvas canvas = new Canvas(result);
+        Paint paint = new Paint();
+        ColorMatrix cm = new ColorMatrix();
+        cm.setSaturation(0f);
+        paint.setColorFilter(new ColorMatrixColorFilter(cm));
+        canvas.drawBitmap(src, 0f, 0f, paint);
+        return result;
+    }
```
[the remark this answers is not in the document]
Comment (deeksha-rgb) [reply clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/Utils.java:617]: Fixed in commit 63008fbcd. Added null and isRecycled() check at the start of toGreyscale(). Returns null for invalid input so callers fall back to the outline drawable.

File: clevertap-pushtemplates/src/main/res/layout/rating.xml [hunk 6]
```diff
@@ -67,18 +68,40 @@
...
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:background="@color/customButtonBackground">
+
+                <ImageView
+                    android:id="@+id/rating_confirm_bg"
+                    android:layout_width="match_parent"
+                    android:layout_height="match_parent"
+                    android:scaleType="fitXY"
+                    android:contentDescription="@string/confirm_btn_content_description" />
```
[the remark this answers is not in the document]
Comment (deeksha-rgb) [reply clevertap-pushtemplates/src/main/res/layout/rating.xml]: Fixed in commit 63008fbcd. Added android:importantForAccessibility="no" to rating_confirm_bg. TalkBack will now only announce the label from tVRatingConfirmation

File: clevertap-pushtemplates/src/test/java/com/clevertap/android/pushtemplates/UtilsTest.kt [hunk 7]
```diff
@@ -856,7 +856,13 @@ class UtilsTest {
...
     fun `getImageDataListFromExtras should return list of ImageData objects`() {
         // Given
-        val keys = setOf("pt_img1", "pt_img2", "other_key", "pt_img", "pt_img_alt_text", "pt_img1_alt_text")
+        // Only pt_img<digit> keys are carousel images. Bare "pt_img" and the pt_img_* border
+        // configuration keys must be ignored even though they all share the "pt_img" prefix.
+        val keys = setOf(
+            "pt_img1", "pt_img2", "other_key", "pt_img", "pt_img_alt_text", "pt_img1_alt_text",
+            PTConstants.PT_IMG_BORDER_CLR, PTConstants.PT_IMG_CORNER_RADIUS,
+            PTConstants.PT_IMG_BORDER_WIDTH
+        )
```
[the remark this answers is not in the document]
Comment (deeksha-rgb) [reply clevertap-pushtemplates/src/test/java/com/clevertap/android/pushtemplates/UtilsTest.kt:865]: Good catch. Will add this edge case to UtilsTest in a follow-up. The production predicate currently handles it correctly — pt_img1_extra ends with _extra not _alt_text so it would be included, which is an existing gap. Will fix the predicate and test together.

File: clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/content/RatingContentView.kt [hunk 8]
```diff
@@ -138,4 +163,41 @@ internal class RatingContentView(
...
+                    NotificationBitmapUtils.createRadialBitmap(c1, c2, 200, 50, borderRadius, borderColor, borderWidth)
+                else null
+            }
+            ButtonStyle.SOLID -> {
+                val bgColor = extras.getString(PTConstants.PT_BTN_CLR)?.let { Utils.getColourOrNull(it) }
+                if (bgColor != null)
+                    NotificationBitmapUtils.createSolidBitmap(bgColor, borderColor, 200, 50, borderRadius, borderWidth)
+                else null
+            }
+        }
```
[the remark this answers is not in the document]
Comment (deeksha-rgb) [reply clevertap-pushtemplates/src/main/java/com/clevertap/android/pushtemplates/content/RatingContentView.kt:196]: Acknowledged. This is a known limitation of the fixed-size bitmap approach with fitXY. A proper fix requires runtime view dimensions which are not available here.

**Comment by deeksha-rgb** [comment]
@coderabbitai review

**Comment by deeksha-rgb** [comment]
@coderabbitai any comments..?
